### ISSUE TITLE
go generate tool

### ISSUE THREAD

**Issue by bmoylan** [palantir/godel#67]
It would be great to have godel support for `go generate`. Unfortunately for purposes of vendoring, generated code should be checked-in instead of generated on-the-fly.

I think it should work similarly to `format` where there are `apply=true` and `apply=false` modes and it is included in `verify`. Whitelist directories in a `godel/config/generate.yml`. `generate` tasks must result in deterministic, stable output.

**Comment by nmiyake** [comment]
+1, generally think this would make sense. Right now, the validation for this is typically done as part of build logic (for example, gödel itself does this as https://github.com/palantir/godel/blob/master/circle.yml#L224), but it's pretty unseemly and inefficient since you also want to be able to run all verifications locally.

**Comment by nmiyake** [comment]
Thought about this a bit more and can think of implementation for all of this. However, the one thing that won't quite fit conceptually is `apply=false`.

The config for this task would consist of multiple `generator` entries, where each `generator` consists of the path to the `go generate` file and a list of path matchers (exact or regexp) that match the output files generated by the task.

Applying the generation is just a matter of running each of the `go generate` generators. In order to figure out whether or not a change was made, we do the following:

* Before running the generator, find all files/directories matched by the matchers for the generator specified in the config and compute hashes for all the files and directories
* Run the generator
* Find all files/directories matched by matchers for the generator, compute hashes, and compare to values computed beforehand
* If there are any differences, then fail verify; otherwise, OK

The issue is that, by the time we have determined that there is a difference, the generator has already run and files have already been modified. Thus, `--apply=false` is not true -- the changes that would have been caused by generation have already happened.

Because of this, it may make sense to have a generate task and a verify option for that task (`./godelw generate` and `./godelw generate --verify-no-changes`), but to not necessarily include it in `verify` itself.

Another option would be to change the `apply=false` flag to be something else like `--verify-no-changes` and make it clear that running this command can actually apply changes to the repository (but is primarily meant for CI environments where such changes can be ignored/rolled back if the task itself fails).
